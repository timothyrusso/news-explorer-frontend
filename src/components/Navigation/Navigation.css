.navigation {
  max-width: 1440px;
  width: 100%;
  border-bottom: 2px solid rgba(196, 196, 196, 0.01);
  box-shadow: inset 0px -1px 0px rgba(255, 255, 255, 0.2);
  min-height: 80px;
  display: flex;
  justify-content: space-between;
  padding: 0 104px;
  position: fixed;
  box-sizing: border-box;
  color: #fff;
}

.navigation__wrapper {
  min-height: 48px;
  display: flex;
  justify-content: space-between;
  align-items: center;
}

.logo {
  min-height: 24px;
  max-width: 200px;
}

.navigation__link {
  font-family: 'Roboto';
  font-style: normal;
  font-weight: 500;
  font-size: 16px;
  line-height: 1.5;
  text-decoration: none;
  color: #fff;
}

.navigation__link:focus {
  border-bottom: 3 px solid #fff;
}

.navigation__button {
  border: 1px solid #fff;
  border-radius: 100px;
  background-color: transparent;
  padding: 12px 25px;
  min-width: 176px;
  color: #fff;
  margin-left: 34px;
}
