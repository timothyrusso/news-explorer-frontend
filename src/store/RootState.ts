import { UserState } from './user/user.reducer';
import { ArticleState } from './article/article.reducer';
import { TogglesState } from './toggles/toggles.reducer';
import { ErrorsState } from './errors/errors.reducer';

export type RootState = {
  user: UserState;
  article: ArticleState;
  toggles: TogglesState;
  errors: ErrorsState;
};
