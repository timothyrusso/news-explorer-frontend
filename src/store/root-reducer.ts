import { combineReducers } from 'redux';
import { userReducer } from './user/user.reducer';
import { articleReducer } from './article/article.reducer';

export const rootReducer = combineReducers({
  user: userReducer,
  article: articleReducer,
});
