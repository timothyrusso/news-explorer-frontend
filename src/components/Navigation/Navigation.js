import './Navigation.css';
import { Link } from 'react-router-dom';
import { useState, useEffect } from 'react';
import { useLocation } from 'react-router-dom';
import { useDispatch } from 'react-redux';
import logo from '../../images/NewsExplorer.svg';
import logout from '../../images/logout.svg';
import blackLogo from '../../images/NewsExplorerlogo-black.svg';
import blackLogout from '../../images/logout-black.svg';
import { useSelector } from 'react-redux';
import { usePopup } from '../../hooks/usePopup';
import { useHandleLogout } from '../../hooks/useHandleLogout';
import { setIsMobileNavbarOppositeAction } from '../../store/toggles/toggles.actions';

const Navigation = () => {
  const [navbarColor, setNavbarColor] = useState(false);
  const [screenWidth, setScreenWidth] = useState(window.innerWidth);

  const currentUser = useSelector((state) => state.user.currentUser);
  const isMobileNavbar = useSelector((state) => state.toggles.isMobileNavbar);

  const { handleSigninPopupClick } = usePopup();
  const { handleLogout } = useHandleLogout();

  const dispatch = useDispatch();

  const isLoggedIn = useSelector((state) => state.toggles.isLoggedin);
  const isBlackNavbar = useSelector((state) => state.toggles.isBlackNavbar);
  const toggleMenu = useSelector((state) => state.toggles.isMobileNavbar);

  const toggleNav = () => {
    dispatch(setIsMobileNavbarOppositeAction(!isMobileNavbar));
  };

  const location = useLocation();

  const colorRuleDefiner =
    (isBlackNavbar && !toggleMenu) || (screenWidth > 700 && isBlackNavbar);

  const buttonColor = colorRuleDefiner
    ? { color: 'black', borderColor: 'black' }
    : { color: 'white', borderColor: 'white' };

  //navbar scroll changeBackground function
  const changeBackground = () => {
    if (window.scrollY >= 66) {
      setNavbarColor(true);
    } else {
      setNavbarColor(false);
    }
  };

  useEffect(() => {
    changeBackground();
    // adding the event when scroll change background
    window.addEventListener('scroll', changeBackground);
    return () => window.removeEventListener('scroll', changeBackground);
  }, []);

  useEffect(() => {
    const changeWidth = () => {
      setScreenWidth(window.innerWidth);
    };
    window.addEventListener('resize', changeWidth);
    return () => window.removeEventListener('resize', changeWidth);
  }, []);

  return (
    <section
      className={`navigation ${
        toggleMenu && screenWidth < 700 ? 'navigation_type_mobile' : ''
      }`}
    >
      <nav
        className={`navigation__bar ${
          navbarColor ? 'navigation__bar_type_scroll' : ''
        }`}
        style={
          toggleMenu && screenWidth < 700 ? { backgroundColor: '#1A1B22' } : {}
        }
      >
        <Link to={'/'} className="logo">
          <img
            src={colorRuleDefiner ? blackLogo : logo}
            alt="NewsExplorer logo"
          ></img>
        </Link>
        {(toggleMenu || screenWidth > 700) && (
          <ul className="navigation__links-wrapper">
            <Link
              to={'/'}
              className={`navigation__link               ${
                location.pathname === '/' ? 'navigation__link_type_focus' : ''
              }`}
              style={buttonColor}
            >
              Home
            </Link>
            <Link
              to={'/saved-news'}
              className={`navigation__link navigation__link_type_saved ${
                !isLoggedIn ? 'navigation__disabled' : ''
              }
              ${
                location.pathname === '/saved-news'
                  ? 'navigation__link_type_focus'
                  : ''
              }
              `}
              style={buttonColor}
            >
              Saved articles
            </Link>
            <button
              aria-label="signin"
              type="button"
              className={`navigation__button navigation__button_type_signin ${
                isLoggedIn ? 'navigation__disabled' : ''
              }`}
              onClick={handleSigninPopupClick}
              style={buttonColor}
            >
              Sign in
            </button>
            <button
              aria-label="signout"
              type="button"
              className={`navigation__button navigation__button_type_logout ${
                !isLoggedIn ? 'navigation__disabled' : ''
              }`}
              onClick={handleLogout}
              style={buttonColor}
            >
              {currentUser.name}
              <img
                src={isBlackNavbar && !toggleMenu ? blackLogout : logout}
                alt="Logout logo"
                className="navigation__logout-icon"
              ></img>
            </button>
          </ul>
        )}
        <button
          onClick={toggleNav}
          className={`burgher-button
          ${isBlackNavbar && !toggleMenu && 'burgher-button_type_black'}
          ${
            toggleMenu
              ? 'burgher-button_type_open'
              : 'burgher-button_type_white'
          }`}
        ></button>
      </nav>
    </section>
  );
};

export default Navigation;
