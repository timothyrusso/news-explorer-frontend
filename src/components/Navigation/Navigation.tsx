import React from 'react';
import { Link } from 'react-router-dom';
import { useLocation } from 'react-router-dom';
import { useDispatch } from 'react-redux';
import logo from '../../images/NewsExplorer.svg';
import logout from '../../images/logout.svg';
import blackLogo from '../../images/NewsExplorerlogo-black.svg';
import blackLogout from '../../images/logout-black.svg';
import { usePopup } from '../../hooks/usePopup';
import { useHandleLogout } from '../../hooks/useHandleLogout';
import { setIsMobileNavbarOppositeAction } from '../../store/toggles/toggles.actions';
import { useAppSelector } from '../../hooks/useAppSelector';
import { useScrollBehavior } from '../../hooks/useScrollBehavior';
import { useWindowResize } from '../../hooks/useWindowResize';
import './Navigation.css';

const Navigation = () => {
  const currentUser = useAppSelector((state) => state.user.currentUser);
  const isMobileNavbar = useAppSelector(
    (state) => state.toggles.isMobileNavbar
  );
  const isLoggedIn = useAppSelector((state) => state.toggles.isLoggedin);
  const isBlackNavbar = useAppSelector((state) => state.toggles.isBlackNavbar);
  const toggleMenu = useAppSelector((state) => state.toggles.isMobileNavbar);

  const { handleSigninPopupClick } = usePopup();
  const { handleLogout } = useHandleLogout();
  const { navbarColor } = useScrollBehavior();
  const { screenWidth } = useWindowResize();

  const dispatch = useDispatch();

  const toggleNav = () => {
    dispatch(setIsMobileNavbarOppositeAction(!isMobileNavbar));
  };

  const location = useLocation();

  const colorRuleDefiner =
    (isBlackNavbar && !toggleMenu) || (screenWidth > 700 && isBlackNavbar);

  const buttonColor = colorRuleDefiner
    ? { color: 'black', borderColor: 'black' }
    : { color: 'white', borderColor: 'white' };

  return (
    <section
      className={`navigation ${
        toggleMenu && screenWidth < 700 ? 'navigation_type_mobile' : ''
      }`}
    >
      <nav
        className={`navigation__bar ${
          navbarColor ? 'navigation__bar_type_scroll' : ''
        }`}
        style={
          toggleMenu && screenWidth < 700 ? { backgroundColor: '#1A1B22' } : {}
        }
      >
        <Link to={'/'} className="logo">
          <img
            src={colorRuleDefiner ? blackLogo : logo}
            alt="NewsExplorer logo"
          ></img>
        </Link>
        {(toggleMenu || screenWidth > 700) && (
          <ul className="navigation__links-wrapper">
            <Link
              to={'/'}
              className={`navigation__link               ${
                location.pathname === '/' ? 'navigation__link_type_focus' : ''
              }`}
              style={buttonColor}
            >
              Home
            </Link>
            <Link
              to={'/saved-news'}
              className={`navigation__link navigation__link_type_saved ${
                !isLoggedIn ? 'navigation__disabled' : ''
              }
              ${
                location.pathname === '/saved-news'
                  ? 'navigation__link_type_focus'
                  : ''
              }
              `}
              style={buttonColor}
            >
              Saved articles
            </Link>
            <button
              aria-label="signin"
              type="button"
              className={`navigation__button navigation__button_type_signin ${
                isLoggedIn ? 'navigation__disabled' : ''
              }`}
              onClick={handleSigninPopupClick}
              style={buttonColor}
            >
              Sign in
            </button>
            <button
              aria-label="signout"
              type="button"
              className={`navigation__button navigation__button_type_logout ${
                !isLoggedIn ? 'navigation__disabled' : ''
              }`}
              onClick={handleLogout}
              style={buttonColor}
            >
              {'__v' in currentUser && currentUser.name}
              <img
                src={isBlackNavbar && !toggleMenu ? blackLogout : logout}
                alt="Logout logo"
                className="navigation__logout-icon"
              ></img>
            </button>
          </ul>
        )}
        <button
          onClick={toggleNav}
          className={`burgher-button
          ${isBlackNavbar && !toggleMenu && 'burgher-button_type_black'}
          ${
            toggleMenu
              ? 'burgher-button_type_open'
              : 'burgher-button_type_white'
          }`}
        ></button>
      </nav>
    </section>
  );
};

export default Navigation;
