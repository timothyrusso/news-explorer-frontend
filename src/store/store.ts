import { legacy_createStore as createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import { rootReducer } from './root-reducer';
import { ThunkDispatch } from 'redux-thunk';
import { AnyAction } from 'redux';
import { RootState } from './RootState';

const store = createStore(rootReducer, applyMiddleware(thunk));

export type AppDispatch = ThunkDispatch<RootState, unknown, AnyAction>;

export default store;
