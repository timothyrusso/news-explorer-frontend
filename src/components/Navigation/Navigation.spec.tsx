import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import store from '../../store/store';
import { BrowserRouter as Router } from 'react-router-dom';
import Navigation from './Navigation';

describe('Navigation component', () => {
  it('Should render the Navigation component', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <Router>
          <Navigation />
        </Router>
      </Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
