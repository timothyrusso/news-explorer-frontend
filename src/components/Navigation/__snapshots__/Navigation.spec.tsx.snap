// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation component Should render the Navigation component 1`] = `
<DocumentFragment>
  <section
    class="navigation "
  >
    <nav
      class="navigation__bar "
    >
      <a
        class="logo"
        href="/"
      >
        <img
          alt="NewsExplorer logo"
          src="test-file-stub"
        />
      </a>
      <ul
        class="navigation__links-wrapper"
      >
        <a
          class="navigation__link               navigation__link_type_focus"
          href="/"
          style="color: white; border-color: white;"
        >
          Home
        </a>
        <a
          class="navigation__link navigation__link_type_saved navigation__disabled
              
              "
          href="/saved-news"
          style="color: white; border-color: white;"
        >
          Saved articles
        </a>
        <button
          aria-label="signin"
          class="navigation__button navigation__button_type_signin "
          style="color: white; border-color: white;"
          type="button"
        >
          Sign in
        </button>
        <button
          aria-label="signout"
          class="navigation__button navigation__button_type_logout navigation__disabled"
          style="color: white; border-color: white;"
          type="button"
        >
          <img
            alt="Logout logo"
            class="navigation__logout-icon"
            src="test-file-stub"
          />
        </button>
      </ul>
      <button
        class="burgher-button
          false
          burgher-button_type_white"
      />
    </nav>
  </section>
</DocumentFragment>
`;
